lib/resource_waiter.js

```diff
@@ -32,76 +61,72 @@ AWS.ResourceWaiter = inherit({
-          resp.error.retryDelay = waiter.config.interval * 1000;
+  matchers: {
+    path: function(resp, expected, argument) {
+      var result = jmespath.search(resp.data, argument);
+      return jmespath.strictDeepEqual(result,expected);
+    },
+
+    pathAll: function(resp, expected, argument) {
+      var results = jmespath.search(resp.data, argument);
+      if (!Array.isArray(results)) results = [results];
+      var numResults = results.length;
+      if (!numResults) return false;
+      for (var ind = 0 ; ind < numResults; ind++) {
+        if (!jmespath.strictDeepEqual(results[ind], expected)) {
+          return false;
         }
-      });
-    }),
-
-    output: new AWS.SequentialExecutor().addNamedListeners(function(add) {
-      add('CHECK_OUT_ERROR', 'extractError', function CHECK_OUT_ERROR(resp) {
-        if (resp.error) {
-          resp.request._waiter.setError(resp, true);
+      }
+      return true;
+    },
+
+    pathAny: function(resp, expected, argument) {
+      var results = jmespath.search(resp.data, argument);
+      if (!Array.isArray(results)) results = [results];
+      var numResults = results.length;
+      for (var ind = 0 ; ind < numResults; ind++) {
+        if (jmespath.strictDeepEqual(results[ind], expected)) {
+          return true;
         }
-      });
+      }
+      return false;
+    },
+
+    status: function(resp, expected) {
+      var statusCode = resp.httpResponse.statusCode;
+      return (typeof statusCode === 'number') && (statusCode === expected);
+    },
+
+    error: function(resp, expected) {
+      if (typeof expected === 'string' && resp.error) {
+        return expected === resp.error.code;
+      }
+      // if expected is not string, can be boolean indicating presence of error
+      return expected === !!resp.error;
+    }
+  },
 
-      add('CHECK_OUTPUT', 'extractData', function CHECK_OUTPUT(resp) {
-        var waiter = resp.request._waiter;
-        var success = waiter.checkSuccess(resp);
-        if (!success) {
-          waiter.setError(resp, success === null ? false : true);
-        } else {
-          resp.error = null;
-        }
-      });
-    }),
-
-    error: new AWS.SequentialExecutor().addNamedListeners(function(add) {
-      add('CHECK_ERROR', 'extractError', function CHECK_ERROR(resp) {
-        var waiter = resp.request._waiter;
-        var success = waiter.checkError(resp);
-        if (!success) {
-          waiter.setError(resp, success === null ? false : true);
-        } else {
-          resp.error = null;
-          resp.data = {};
-          resp.request.removeAllListeners('extractData');
-        }
-      });
+  listeners: new AWS.SequentialExecutor().addNamedListeners(function(add) {
+    add('RETRY_CHECK', 'retry', function(resp) {
+      var waiter = resp.request._waiter;
+      if (resp.error && resp.error.code === 'ResourceNotReady') {
+        resp.error.retryDelay = (waiter.config.delay || 0) * 1000;
```

Comment:
It might make sense to use a different default for the delay than 0, maybe 1. This is an edge-case because I think all our waiters have a delay specified, but we would potentially see throttling errors if a delay of 0 was used.